python/pyspark/sql/tests/test_arrow.py

```diff
@@ -190,6 +190,13 @@ def test_pandas_round_trip(self):
         pdf_arrow = df.toPandas()
         assert_frame_equal(pdf_arrow, pdf)
 
+    def test_pandas_self_destruct(self):
+        with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": True}):
+            pdf = self.create_pandas_data_frame()
```

Comment:
This is a pretty minimal dataframe, mostly for checking type compatibility. It would be nice to also check against a larger dataframe that the memory usage doesn't double during conversion, but I'm not sure how to do that reliably.

Reply:
It's really hard to verify this in an automated way due to how jemalloc works (except possibly by reconfiguring jemalloc) - at the very least it'd be iffy to do in a unit test like this without some setup.

(The brief explanation is that with how Arrow configures jemalloc, memory is only slowly reclaimed by the OS, so in naive measurements, you'll still see a memory usage spike - albeit slightly less of one. In C++, you can force jemalloc to purge everything right away, but even via ctypes I couldn't access the right functions from Python.)

Reply:
Well, I was thinking you might be able to use `pyarrow.total_allocated_bytes()` - this is how it's tested in pyarrow https://github.com/apache/arrow/blob/e6eb61f58ef382003c9462924563f575d9a59c13/python/pyarrow/tests/test_pandas.py#L3340. But the table is deallocated after `toPandas()` returns anyway, so you would have to sample before and after `pdf = table.to_pandas(**pandas_options)`..

Reply:
Hmm, is that test quite right? self_destruct doesn't change the end memory consumption, but rather the peak consumption...which you can't quite reliably measure from Python (via a separate thread, perhaps?).

Reply:
The pyarrow test works because it allocates the memory for the required Pandas blocks too, which I believe doubles the memory allocated normally, but ends up with a net zero change with self destruct.

Reply:
Sorry again for the delay.

With some refactoring, this test now functions similarly to the PyArrow test. However if we don't want to rearrange things like this I can revert this change.